x-pack/plugin/security/lib/nimbus-jose-jwt-fixed/src/main/java/com/nimbusds/jose/util/JSONObjectUtils.java

```diff
@@ -0,0 +1,519 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package com.nimbusds.jose.util;
+
+import com.nimbusds.jose.shaded.gson.Gson;
+import com.nimbusds.jose.shaded.gson.GsonBuilder;
+import com.nimbusds.jose.shaded.gson.ToNumberPolicy;
+import com.nimbusds.jose.shaded.gson.reflect.TypeToken;
+
+import java.lang.reflect.Type;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.text.ParseException;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Copied from nimbus-jose-jwt version 9.37.3.
+ *
+ * Original code Copyright 2012-2016, Connect2id Ltd. Licensed under the Apache License, Version 2.0
+ *
+ * The only modifications in this file are:
+ * 1) {@link AccessController#doPrivileged(PrivilegedAction)} calls to make gson work with the security manager
+ * 2) Formatting/Warning suppression as necessary to work with our infrastructure
+ * 3) This comment and the license comment
+ */
+@SuppressWarnings({ "unchecked", "rawtypes" })
+public class JSONObjectUtils {
```

Comment:
I understand the desire to avoid lots of doPrivileged calls. However, this seems quite a bit more fragile than other times we have replaced classes from dependencies. For example, with the hdfs client we replaced the shutdown manager with noop methods. In contrast, here many method implementations are copied, and if in future updates they are modified, we might still have the old implementation. It seems like a bigger risk, I lean towards having the doPrivileged calls (though admittedly I did not look at the previous PR iteration, so I don't know how pervasive the doPrivs were).

Reply:
It's less that the doPriv calls were pervasive and more that a *lot* of things end up calling into one of these methods, including constructors and `toString` methods, to the point where it's not just "pervasive calls", it's "it's legit really hard to add any new code without breaking something".

To take an example @n1v0lg provided (thanks!), consider these two lines:
```
logger.info("Claim set small whoops: {}", jwtAuthenticationToken.getJWTClaimsSet());
logger.info("Claim set big whoops: " + jwtAuthenticationToken.getJWTClaimsSet());
```

Small whoops ends up logging a message like:
`Claim set small whoops: [!!!com.nimbusds.jwt.JWTClaimsSet@15ee0d6d=>java.security.AccessControlException:access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")!!!]`

Big whoops throws an exception. It's very, very, very easy to accidentally hit this, and not even necessarily cause a test failure about it. While we can work around that with forbiddenAPIs (which is what I started doing and can revert back to if we would really, *really* rather do it that way), it's a huge pain in the butt now and any time we have to do anything with the JWT or OIDC realms going forward. On top of that, the forbiddenAPIs list is made by me going through the libraries in question with IntelliJ, tracing all the various ways that `JSONObjectUtils` can be called into. This would need to be updated manually every time we update these dependencies, and less obviously so - at least this is likely to trigger a compilation failure if there are significant changes, whereas a forbiddenAPIs-based solution not being properly updated when the dependency would result in silently permitting us to ship one or both of the "whoops" cases above.

All the options suck. I'm open to being talked into the idea that the forbiddenAPIs-based solution sucks less than vendoring, but it's not good.

Reply:
I generally like this approach better. However, rather than removing the original class then replacing functionality with wrap(copy/paste), can we instead move the original class to different package then replace the functionality with wrap(delegate(method)) ?

So instead of duplicating the source code, we keep their original source but in a different package name and just wrap the calls to the original code inside the doPrivilged ? It should surface any Java API breaking changes to these methods as compile time issues the next version bump. 

This won't help with the 2 private static method/members, but those seem unlikely to change and we if we really want to (i don't think it is necessary) we could use the elevated privs to delegate to them too via reflections. It might worth a unit test or two to use reflections and mocks to help ensure the same behavior between the original and what is copied, or maybe just a comment in the build.gradle to double check those on a version bump.

Reply:
Ooh, I like that idea, I think it addresses the worst of the concerns Ryan raised. I'll get that spun together, thanks.

Reply:
@jakelandis I've put together an approach with `JSONObjectUtils` (and `JSONStringUtils`) copied verbatim (sans formatting) into `InnerJSONObjectUtils` and produced wrappers under the original names. It's not quite as clean as it maybe could be since everything is static, but I think it's still an improvement.